pkg/snapshot/clonetree_pci.go

```diff
@@ -0,0 +1,135 @@
+// ghw cares about pertaining to all the PCI devices connected to the bus connected from the
+// given root; usually (but not always) a CPU packages has 1+ PCI(e) roots, forming the first
+// level; more PCI bridges are (usually) attached to this level, creating deep nested trees.
+// hence we need to scan all possible roots, to make sure not to miss important devices.
+func scanPCIDeviceRoot(root string) (fileSpecs []string, pciRoots []string) {
+	trace("scanning PCI device root %q\n", root)
+
+	perDevEntries := []string{
+		"class",
+		"device",
+		"irq",
+		"local_cpulist",
+		"modalias",
+		"numa_node",
+		"revision",
+		"vendor",
+	}
+	entries, err := ioutil.ReadDir(root)
+	if err != nil {
+		return []string{}, []string{}
+	}
+	for _, entry := range entries {
+		entryName := entry.Name()
+		if addr := pciaddr.FromString(entryName); addr == nil {
+			// doesn't look like a entry we care about
+			continue
+		}
+
+		entryPath := filepath.Join(root, entryName)
+		pciEntry := findPCIEntryFromPath(root, entryName)
+		trace("PCI entry is %q\n", pciEntry)
+		fileSpecs = append(fileSpecs, entryPath)
+		for _, perNetEntry := range perDevEntries {
+			fileSpecs = append(fileSpecs, filepath.Join(pciEntry, perNetEntry))
+		}
+
+		if isPCIBridge(entryPath) {
+			trace("adding new PCI root %q\n", entryName)
+			pciRoots = append(pciRoots, pciEntry)
+		}
+	}
+	return fileSpecs, pciRoots
+}
+
+func findPCIEntryFromPath(root, entryName string) string {
+	entryPath := filepath.Join(root, entryName)
+	fi, err := os.Lstat(entryPath)
+	if err != nil {
+		trace("stat(%s) failed: %v\n", entryPath, err)
+		return entryPath // what else we can do?
+	}
+	if fi.Mode()&os.ModeSymlink == 0 {
+		// regular file, nothing to resolve
+		return entryPath
+	}
+	// resolve symlink
+	target, err := os.Readlink(entryPath)
+	trace("entry %q is symlink resolved to %q\n", entryPath, target)
+	if err != nil {
+		trace("readlink(%s) failed: %v - skipped\n", entryPath, err)
+		return entryPath // what else we can do?
+	}
+	return filepath.Clean(filepath.Join(root, target))
+}
+
+func isPCIBridge(entryPath string) bool {
+	subNodes, err := ioutil.ReadDir(entryPath)
+	if err != nil {
+		trace("error scanning device entry path %q: %v", entryPath, err)
+		return false
+	}
+	for _, subNode := range subNodes {
+		if !subNode.IsDir() {
+			continue
+		}
+		if addr := pciaddr.FromString(subNode.Name()); addr != nil {
+			// we got an entry in the directory pertaining to this device
+			// which is a directory itself and it is named like a PCI address.
+			// Hence we infer the device we are considering is a PCI bridge of sorts.
+			// This is is indeed a bit brutal, but the only possible alternative
+			// (besides blindly copying everything in /sys/bus/pci/devices) is
+			// to detect the type of the device and pick only the bridges.
+			// This approach duplicates the logic with the `pci` subkpg
+			// - or forces us into awkward dep cycles, and has poorer forward
+			// compatibility.
```

Comment:
good explanation. :+1:

Reply:
thanks. The original implementation duplicated code from `pci` package to detect the pci bridges. This is probably more conceptually like the right thing to do, but in practice is less maintenable.

Reply:
also: we cannot just consume the `pci` package here (hence the `pciaddress` split!) again because of the circular dep to `context`.